### PR TITLE
UCS/SYS: Improve IO error handling

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Improve IO error handling

## Why ?

This is needed for TCP #3547 

## How ?

Implemented `ucs_socket_io_err_info_t` structure that is passed to error handling callback

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
> what does this address/improve?

discussed offline - removed calling callback from another callback

@brminich can you review? is it ok now?

File: src/ucs/sys/sock.c [hunk 1]
```diff
@@ -178,6 +178,44 @@ int ucs_socket_max_conn()
...
+        .err_no    = err_no
+    };
+
+    if (err_cb != NULL) {
+        if (err_cb(err_cb_arg, &err_info) == UCS_OK) {
+            return;
+        }
+    }
+
+    ucs_socket_io_print_error(&err_info);
```
Comment (brminich) [inline src/ucs/sys/sock.c]: you could print error here, then no need to have `ucs_socket_io_err_info_t`

**Comment by dmitrygladkov** [comment]
> imo, looks unnecessary/complicated
> 
> maybe just improve/extend error message in `ucs_socket_do_io_nb`?

@brminich I've updated the PR to do it simpler and w/ needed functionality
force-pushed - it was replaced by the new implementation

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10194/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10201/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_cm.c [hunk 2]
```diff
@@ -60,16 +60,20 @@ void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep,
                                str_remote_addr, UCS_SOCKADDR_STRING_LEN));
 }
 
-static void uct_tcp_cm_io_err_handler_cb(void *arg, int errno)
+static ucs_status_t
+uct_tcp_cm_io_err_handler_cb(void *arg, int err_no)
```
Comment (brminich) [inline src/uct/tcp/tcp_cm.c]: minor: better used `errno` here and in `uct_tcp_ep_io_err_handler_cb`, because it is well known/used name

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10212/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_cm.c (hunk 2 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp_cm.c]: fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10216/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe could you review pls?

**Comment by yosefe** [comment]
@dmitrygladkov can we now add UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE capability flag to uct_tcp_iface_query()

**Comment by dmitrygladkov** [comment]
> can we now add UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE capability flag to uct_tcp_iface_query()

yes, we can do it. but let's add it after full TCP connection deduplication will be merged to test it with those changes, ok?

**Pull request merged**